In [135]:
import numpy as np
import pandas as pd
from pandas import read_csv
from pandas import DataFrame
from keras.models import Sequential
from keras.layers import Dense
from keras.wrappers.scikit_learn import KerasClassifier
from keras.utils import np_utils
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import KFold
from sklearn.preprocessing import LabelEncoder
from sklearn.pipeline import Pipeline
from astropy.table import Table
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import StratifiedKFold

We will train a model to automatically classify galaxies by morphology (labeled data from SDSS and Galaxy Zoo). 
We follow Banjeri et al 2010, MNRAS 406, 342-353

In [136]:
file = 'data/Skyserver_galaxymorph2.fit'
data = Table.read(file, format='fits')
df = data.to_pandas()
df.head()

dered_g    dered_r    dered_i   deVAB_i   expAB_i     lnLexp_i  \
0  16.157143  15.339664  14.977295  0.771919  0.731467 -9126.769531   
1  17.727175  17.277601  17.138424  0.718036  0.752297    -0.122561   
2  17.465136  17.251240  17.083557  0.788438  0.828909   -59.886116   
3  16.852707  16.538921  16.299740  0.292246  0.219601   -14.584803   
4  17.990713  17.596178  17.506077  0.193646  0.842761  -100.243294   

     lnLdeV_i     lnLstar_i  nvote   p_el   p_cw  p_acw  p_edge   p_dk   p_mg  \
0 -393.420349 -42334.523438     21  1.000  0.000  0.000   0.000  0.000  0.000   
1  -22.863409   -321.968689     30  0.167  0.133  0.433   0.167  0.100  0.000   
2  -60.950695  -1870.144775     30  0.400  0.000  0.133   0.267  0.167  0.033   
3 -435.035919  -2757.247070     33  0.061  0.000  0.000   0.909  0.000  0.030   
4  -71.750786   -104.684471     66  0.364  0.000  0.015   0.030  0.348  0.242   

    p_cs  
0  0.000  
1  0.733  
2  0.400  
3  0.909  
4  0.045

In [137]:
# fix random seed for reprodutable = Table.read(data)
seed = 7
numpy.random.seed(seed)

In [138]:
#df['g-r', 'r-i'] = df['dered_g']-df['dered_r'], df['dered_r']-df['dered_i']
df['g-r']= df['dered_g']-df['dered_r']
df['r-i']= df['dered_r']-df['dered_i']
df.head()

dered_g    dered_r    dered_i   deVAB_i   expAB_i     lnLexp_i  \
0  16.157143  15.339664  14.977295  0.771919  0.731467 -9126.769531   
1  17.727175  17.277601  17.138424  0.718036  0.752297    -0.122561   
2  17.465136  17.251240  17.083557  0.788438  0.828909   -59.886116   
3  16.852707  16.538921  16.299740  0.292246  0.219601   -14.584803   
4  17.990713  17.596178  17.506077  0.193646  0.842761  -100.243294   

     lnLdeV_i     lnLstar_i  nvote   p_el   p_cw  p_acw  p_edge   p_dk   p_mg  \
0 -393.420349 -42334.523438     21  1.000  0.000  0.000   0.000  0.000  0.000   
1  -22.863409   -321.968689     30  0.167  0.133  0.433   0.167  0.100  0.000   
2  -60.950695  -1870.144775     30  0.400  0.000  0.133   0.267  0.167  0.033   
3 -435.035919  -2757.247070     33  0.061  0.000  0.000   0.909  0.000  0.030   
4  -71.750786   -104.684471     66  0.364  0.000  0.015   0.030  0.348  0.242   

    p_cs       g-r       r-i  
0  0.000  0.817478  0.362370  
1  0.733  0.449574  0.139177  
2  0.400  0.213896  0.167683  
3  0.909  0.313786  0.239182  
4  0.045  0.394535  0.090101

In [139]:
#predictors = ['modelMag_u', 'modelMag_g', 'modelMag_r', 'modelMag_i']
predictors = ['g-r', 'r-i', 'deVAB_i', 'expAB_i', 'lnLexp_i', 'lnLdeV_i', 'lnLstar_i']
types = ['p_el', 'p_cs', 'p_acw', 'p_edge', 'p_dk', 'p_mg', 'p_cs']
df['class'] = DataFrame.idxmax(df[types], axis=1)
df.head()

dered_g    dered_r    dered_i   deVAB_i   expAB_i     lnLexp_i  \
0  16.157143  15.339664  14.977295  0.771919  0.731467 -9126.769531   
1  17.727175  17.277601  17.138424  0.718036  0.752297    -0.122561   
2  17.465136  17.251240  17.083557  0.788438  0.828909   -59.886116   
3  16.852707  16.538921  16.299740  0.292246  0.219601   -14.584803   
4  17.990713  17.596178  17.506077  0.193646  0.842761  -100.243294   

     lnLdeV_i     lnLstar_i  nvote   p_el   p_cw  p_acw  p_edge   p_dk   p_mg  \
0 -393.420349 -42334.523438     21  1.000  0.000  0.000   0.000  0.000  0.000   
1  -22.863409   -321.968689     30  0.167  0.133  0.433   0.167  0.100  0.000   
2  -60.950695  -1870.144775     30  0.400  0.000  0.133   0.267  0.167  0.033   
3 -435.035919  -2757.247070     33  0.061  0.000  0.000   0.909  0.000  0.030   
4  -71.750786   -104.684471     66  0.364  0.000  0.015   0.030  0.348  0.242   

    p_cs       g-r       r-i class  
0  0.000  0.817478  0.362370  p_el  
1  0.733  0.449574  0.139177  p_cs  
2  0.400  0.213896  0.167683  p_el  
3  0.909  0.313786  0.239182  p_cs  
4  0.045  0.394535  0.090101  p_el

In [140]:
list(enumerate(types))

[(0, 'p_el'),
 (1, 'p_cs'),
 (2, 'p_acw'),
 (3, 'p_edge'),
 (4, 'p_dk'),
 (5, 'p_mg'),
 (6, 'p_cs')]

In [141]:
#predictors = ['modelMag_u', 'modelMag_g', 'modelMag_r', 'modelMag_i']

In [142]:
df['response'] = 1.
for i in range(len(types)):
    response = types[i]
    print(response)
    msk = df['class'] == response
    df['response'][msk] = i

df[['class', 'response']]   

p_el
p_cs
p_acw


/Users/ceciliagarraffo/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


p_edge
p_dk
p_mg
p_cs


class  response
0      p_el       0.0
1      p_cs       6.0
2      p_el       0.0
3      p_cs       6.0
4      p_el       0.0
5      p_el       0.0
6      p_el       0.0
7      p_el       0.0
8      p_el       0.0
9      p_dk       4.0
10     p_cs       6.0
11     p_cs       6.0
12     p_el       0.0
13     p_el       0.0
14     p_el       0.0
15     p_el       0.0
16     p_dk       4.0
17     p_dk       4.0
18     p_mg       5.0
19     p_mg       5.0
20     p_el       0.0
21     p_el       0.0
22     p_dk       4.0
23     p_el       0.0
24     p_dk       4.0
25     p_cs       6.0
26     p_cs       6.0
27     p_cs       6.0
28     p_el       0.0
29     p_dk       4.0
...     ...       ...
65487  p_el       0.0
65488  p_el       0.0
65489  p_el       0.0
65490  p_el       0.0
65491  p_cs       6.0
65492  p_dk       4.0
65493  p_cs       6.0
65494  p_cs       6.0
65495  p_el       0.0
65496  p_mg       5.0
65497  p_el       0.0
65498  p_cs       6.0
65499  p_el       0.0
65500  p_cs       6.0
65501  p_cs       6.0
65502  p_el       0.0
65503  p_el       0.0
65504  p_el       0.0
65505  p_el       0.0
65506  p_dk       4.0
65507  p_el       0.0
65508  p_cs       6.0
65509  p_el       0.0
65510  p_el       0.0
65511  p_cs       6.0
65512  p_cs       6.0
65513  p_el       0.0
65514  p_el       0.0
65515  p_el       0.0
65516  p_cs       6.0

[65517 rows x 2 columns]

In [143]:
#df['modelMag_u', 'modelMag_g', 'modelMag_r', 'modelMag_i', 'response'].head()

In [144]:
df.head()

dered_g    dered_r    dered_i   deVAB_i   expAB_i     lnLexp_i  \
0  16.157143  15.339664  14.977295  0.771919  0.731467 -9126.769531   
1  17.727175  17.277601  17.138424  0.718036  0.752297    -0.122561   
2  17.465136  17.251240  17.083557  0.788438  0.828909   -59.886116   
3  16.852707  16.538921  16.299740  0.292246  0.219601   -14.584803   
4  17.990713  17.596178  17.506077  0.193646  0.842761  -100.243294   

     lnLdeV_i     lnLstar_i  nvote   p_el   p_cw  p_acw  p_edge   p_dk   p_mg  \
0 -393.420349 -42334.523438     21  1.000  0.000  0.000   0.000  0.000  0.000   
1  -22.863409   -321.968689     30  0.167  0.133  0.433   0.167  0.100  0.000   
2  -60.950695  -1870.144775     30  0.400  0.000  0.133   0.267  0.167  0.033   
3 -435.035919  -2757.247070     33  0.061  0.000  0.000   0.909  0.000  0.030   
4  -71.750786   -104.684471     66  0.364  0.000  0.015   0.030  0.348  0.242   

    p_cs       g-r       r-i class  response  
0  0.000  0.817478  0.362370  p_el       0.0  
1  0.733  0.449574  0.139177  p_cs       6.0  
2  0.400  0.213896  0.167683  p_el       0.0  
3  0.909  0.313786  0.239182  p_cs       6.0  
4  0.045  0.394535  0.090101  p_el       0.0

Split predictors and response

In [145]:
print(predictors[:], ['response'])
df[predictors[:]]

['g-r', 'r-i', 'deVAB_i', 'expAB_i', 'lnLexp_i', 'lnLdeV_i', 'lnLstar_i'] ['response']


g-r       r-i   deVAB_i   expAB_i      lnLexp_i     lnLdeV_i  \
0      0.817478  0.362370  0.771919  0.731467  -9126.769531  -393.420349   
1      0.449574  0.139177  0.718036  0.752297     -0.122561   -22.863409   
2      0.213896  0.167683  0.788438  0.828909    -59.886116   -60.950695   
3      0.313786  0.239182  0.292246  0.219601    -14.584803  -435.035919   
4      0.394535  0.090101  0.193646  0.842761   -100.243294   -71.750786   
5      0.918608  0.424013  0.621988  0.614841  -1319.880249  -311.409271   
6      0.218086  0.296398  0.470560  0.529419   -205.487717    -4.703940   
7      0.596766  0.414845  0.983522  0.974777   -114.888725   -14.876837   
8      0.366402  0.192118  0.582849  0.528655  -2853.558594 -1013.533142   
9      0.366399  0.089875  0.050000  0.076823   -212.310791  -874.054810   
10     0.833862  0.396680  0.473959  0.454091  -3412.221436  -362.977173   
11     0.880548  0.420074  0.545278  0.510447 -22106.107422 -5106.710938   
12     0.873564  0.416084  0.749955  0.757973  -2725.688721  -239.405594   
13    -0.052023  0.256325  0.800848  0.836074     -4.352443  -140.278641   
14     0.996408  0.474211  0.892817  0.885378  -1851.302002  -482.032959   
15     0.882638  0.421146  0.884521  0.836117 -19535.521484  -866.363525   
16     0.882638  0.421146  0.884521  0.836117 -19535.521484  -866.363525   
17     0.626287  0.310177  0.264678  0.352833   -156.267654   -96.733650   
18     0.998516  0.724092  0.674697  0.586087 -13144.727539 -4171.909180   
19     0.662645  0.340740  0.769602  0.767581  -2145.429443 -1396.632324   
20     0.947935  0.427646  0.926930  0.917380  -7146.671875  -171.968948   
21     0.956839  0.445853  0.926973  0.936056  -4304.468750  -176.262329   
22     0.534099 -0.071562  0.877689  0.996674   -391.707520  -121.786880   
23     0.898067  0.406696  0.934005  0.867105 -29777.109375 -2534.202637   
24     0.455187  0.610115  0.677403  0.684274  -1769.562134  -780.196533   
25     0.787699  0.394629  0.790887  0.672849 -13152.822266 -3023.665527   
26     0.822765  0.396178  0.857200  0.860360 -12752.946289 -6478.416992   
27     0.846888  0.397299  0.435197  0.467603  -3932.681885  -670.440430   
28     0.292496  0.208559  0.522945  0.600015    -35.776737   -24.663347   
29     0.361612  0.282715  0.165767  0.271878   -175.570343   -88.443230   
...         ...       ...       ...       ...           ...          ...   
65487  0.900547  0.409946  0.820852  0.821359   -814.240967  -564.739929   
65488  1.041662  0.430581  0.682620  0.721568  -4973.547363  -363.292603   
65489  0.978539  0.438495  0.781364  0.763340  -9277.552734  -269.084778   
65490  0.898153  0.410149  0.718023  0.712601  -3513.866211   -49.961010   
65491  0.755318  0.436039  0.242391  0.229755  -1217.630005 -5382.714844   
65492 -5.343229  0.409138  0.164554  0.176376      0.000000     0.000000   
65493  0.797641  0.377700  0.520962  0.520596  -9183.910156 -1052.414551   
65494  0.675142  0.347205  0.493088  0.521542  -1898.819458 -1700.032593   
65495  0.888816  0.429920  0.849948  0.814291 -17864.623047 -4003.421631   
65496  0.782937  0.308844  0.712425  0.700791  -2817.357666 -1284.900146   
65497  0.544474  0.350954  0.896037  0.883957    -65.588646  -847.773193   
65498  0.816301  0.402528  0.648769  0.584396  -3864.459473 -1986.932251   
65499  0.490744  0.324312  0.452412  0.486004    -54.779247 -1235.652344   
65500  0.791826  0.422522  0.675772  0.721759  -1057.382080  -832.913574   
65501  0.560578  0.344068  0.789138  0.794030  -1085.109375 -1714.647217   
65502  0.900969  0.425291  0.687286  0.710854  -1803.094849   -78.858719   
65503  0.850789  0.398544  0.859513  0.891340 -19220.166016 -1269.124268   
65504  1.052042  0.527756  0.739356  0.702085  -3732.029297  -756.655640   
65505  0.380651  0.152907  0.050000  0.056870   -301.214539  -122.597923   
65506  0.257710  0.259686  0.598624  0.532134    -61.797924  -491.762421   
65507  0.793043  0.509516  0.598207  0.602865  -3618.776123 

In [146]:
df['response'][:10]

0    0.0
1    6.0
2    0.0
3    6.0
4    0.0
5    0.0
6    0.0
7    0.0
8    0.0
9    4.0
Name: response, dtype: float64

In [147]:
select = np.append(predictors, 'response')
msk = np.random.random(len(df['response'])) >= 0.99
dataset = df[select][msk].values
print(dataset.shape)
X = dataset[:, 0:-1].astype(float)
Y = dataset[:,-1]
print(Y)

(629, 8)
[0. 0. 6. 6. 6. 0. 5. 4. 4. 5. 6. 6. 5. 0. 0. 0. 0. 4. 4. 0. 0. 0. 4. 0.
 4. 0. 4. 4. 0. 6. 4. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 4. 6. 6. 0. 6. 6.
 6. 0. 0. 0. 6. 4. 0. 6. 4. 0. 6. 0. 0. 6. 6. 0. 0. 0. 6. 6. 0. 6. 4. 6.
 6. 4. 6. 0. 6. 4. 0. 0. 6. 0. 6. 0. 0. 0. 4. 6. 4. 4. 0. 0. 6. 0. 0. 0.
 0. 0. 6. 6. 0. 6. 4. 6. 6. 0. 0. 0. 6. 6. 0. 0. 0. 6. 0. 6. 0. 6. 0. 4.
 6. 0. 6. 0. 6. 0. 6. 0. 6. 6. 0. 0. 6. 0. 6. 4. 6. 6. 0. 0. 4. 0. 0. 0.
 6. 0. 6. 0. 6. 4. 0. 0. 6. 0. 0. 6. 0. 6. 6. 0. 0. 0. 6. 6. 0. 6. 0. 6.
 0. 5. 0. 0. 0. 0. 4. 0. 6. 0. 0. 0. 4. 0. 6. 6. 0. 6. 0. 6. 0. 0. 6. 0.
 0. 6. 6. 6. 0. 6. 0. 6. 6. 4. 0. 4. 6. 0. 6. 0. 0. 0. 0. 0. 6. 6. 0. 6.
 6. 0. 0. 0. 0. 6. 6. 6. 6. 0. 6. 6. 6. 5. 4. 6. 6. 0. 0. 0. 0. 6. 4. 0.
 6. 0. 6. 6. 0. 6. 6. 0. 6. 4. 0. 4. 6. 0. 0. 6. 6. 6. 6. 0. 6. 6. 0. 0.
 6. 4. 5. 0. 0. 4. 6. 6. 6. 4. 0. 6. 0. 6. 5. 0. 0. 4. 4. 6. 6. 6. 6. 0.
 0. 6. 6. 6. 6. 0. 0. 6. 0. 0. 0. 0. 0. 6. 0. 0. 4. 0. 0. 4. 0. 4. 0. 0.
 6. 6. 5. 6. 6. 0. 0. 6. 0. 6. 6. 6. 0. 0.

One Hot Encoding of The Classes

In [148]:
inp = X.shape[1] #number of inputs

In [149]:
# encode class values as integers
encoder = LabelEncoder()
encoder.fit(Y)
encoded_Y = encoder.transform(Y)
# convert integers to dummy variables (i.e. one hot encoded)
dummy_y = np_utils.to_categorical(encoded_Y)
print(dummy_y.shape[1])

4


Define neural network: 4 inputs -> [8 hidden nodes] -> 4 outputs

In [160]:
# define baseline model
def baseline_model():
    # create model
    model = Sequential()
    model.add(Dense(8, input_dim=inp, activation='relu')) 
    model.add(Dense(8,  activation='relu')) # Do we need a kernel_initializer? ('uniform' or 'normal'?)
    model.add(Dense(dummy_y.shape[1], activation='softmax')) #because it is cathegorical 
    # Compile model
    model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy']) 
    return model

In [161]:
#estimator = KerasClassifier(build_fn=baseline_model, epochs=200, batch_size=5, verbose=0) # verbose = 0 means de\bugging is turned off
#kfold = KFold(n_splits=10, shuffle=True, random_state=seed)
#results = cross_val_score(estimator, X, dummy_y, cv=kfold)
#print("Accuracy: %.2f%% (%.2f%%)" % (results.mean()*100, results.std()*100))

In [162]:
# evaluate baseline model with standardized dataset
estimators = []
estimators.append(('standardize', StandardScaler()))
estimators.append(('mlp', KerasClassifier(build_fn=baseline_model, epochs=100,
    batch_size=5, verbose=0)))
pipeline = Pipeline(estimators)
kfold = StratifiedKFold(n_splits=10, shuffle=True, random_state=seed)
results = cross_val_score(pipeline, X, encoded_Y, cv=kfold)
print("Standardized: %.2f%% (%.2f%%)" % (results.mean()*100, results.std()*100))

Standardized: 66.68% (5.77%)


Clean data. We only want to keep p_el (elliptical), p_cw+p_acw+p_edge (spiral), and p_dk (¨don´t know¨ or point artifacts). We also keep only the ones with 98% certainly on each category.